src/kubernetes_charm.py

```diff
@@ -80,15 +104,198 @@ def _upgrade(self) -> typing.Optional[kubernetes_upgrade.Upgrade]:
         except upgrade.PeerRelationNotReady:
             pass
 
+    @property
+    def _status(self) -> ops.StatusBase:
+        if self.config.get("expose-external", "false") not in [
+            "false",
+            "nodeport",
+            "loadbalancer",
+        ]:
+            return ops.BlockedStatus("Invalid expose-external config value")
+        if (
+            self._peer_data.get_value(
+                relations.secrets.APP_SCOPE, self._K8S_SERVICE_INITIALIZED_KEY
+            )
+            and not self._check_service_connectivity()
+        ):
+            if self._peer_data.get_value(
+                relations.secrets.APP_SCOPE, self._K8S_SERVICE_CREATING_KEY
+            ):
+                return ops.MaintenanceStatus("Waiting for K8s service connectivity")
+            else:
+                return ops.BlockedStatus("K8s service not connectible")
+
     def is_externally_accessible(self, *, event) -> typing.Optional[bool]:
         """No-op since this charm is exposed with node-port"""
 
-    def _reconcile_node_port(self, *, event) -> None:
-        self._patch_service(event)
+    def _get_service(self) -> typing.Optional[lightkube.resources.core_v1.Service]:
+        """Get the managed k8s service."""
+        try:
+            service = self._lightkube_client.get(
+                res=lightkube.resources.core_v1.Service,
+                name=self.service_name,
+                namespace=self.model.name,
+            )
+        except lightkube.core.exceptions.ApiError as e:
+            if e.status.code == 404:
+                return None
+            raise
+
+        return service
+
+    @functools.cache
+    def _get_pod(self, unit_name: str) -> lightkube.resources.core_v1.Pod:
```

Comment:
Are we sure we could cache the result of this function? I am not worried about the _size_ of the cache, but the _validity_ of the result being cached.

Are we sure that, during the execution of an event hook, there can not be multiple units with the same name? If such scenario is imposible, shouldn't the argument be called `unit_id` instead? Not sure if that jargon is used somewhere else.

Reply:
All clear. Unit names are preserve uniqueness.